Result

dpotepalov opened the pull request.

dpotepalov committed:
Move endianness-related code to a separate file

To be reused in binary deserialization

dpotepalov committed:
Implement libpq result parsing in binary format for ints and text

thed636 commented on include/ozo/impl/row.h:
```diff
@@ -0,0 +1,105 @@
+convert_row(RowData&& row_data, const OidMap& oid_map, Row& row) {
+    error_code ec = error::ok;
+    auto data_it = begin(row_data);
+    auto data_end = end(row_data);
```
I think you can make the dimension check here instead of checks in for_each. The Row must be a Sequence, so you can get a size of the sequence and compare it with std::distance between data_it and data_end.

dpotepalov replied:
done

thed636 commented on include/ozo/binary_deserialization.h:
```diff
@@ -0,0 +1,36 @@
+namespace ozo {
+
+inline error_code recv(oid_t, const char* bytes, std::size_t size, std::string& value) {
+    // TODO: client to server encoding conversion?
```
What do you mean?

dpotepalov replied:
I misinterpreted text's recv() code in postgres, it also does encoding conversions on strings, so I thought that maybe we should do the same. We should not, it's done by the server. I'll remove the comment.

thed636 commented on include/ozo/impl/result.h:
```diff
@@ -0,0 +1,49 @@
+
+template <typename Rows, typename OidMap, typename Iterator>
+Require<Iterable<Rows> && (ForwardIterator<Iterator> || InsertIterator<Iterator>), error_code>
+convert_result(Rows&& rows, const OidMap& oid_map, Iterator result) {
```
IMO it is better to use ```out``` name for output result to avoid using a part of function name as a parameter name.

dpotepalov replied:
done

thed636 commented on include/ozo/impl/result.h:
```diff
@@ -0,0 +1,49 @@
+        } else if constexpr (InsertIterator<Iterator>) {
+            using row_value_t = typename Iterator::container_type::value_type;
+
+            auto result_row = row_value_t{};
```
I think it is better to use an overloaded function instead of ```constexpr``` because even with ForwardIterator and not a default constructible type you still can get here a compile error here.

dpotepalov replied:
done

thed636 commented on include/ozo/impl/result.h:
```diff
@@ -0,0 +1,49 @@
+    result(rows_t&& rows) : rows_(std::move(rows)) {}
+
+    template <typename Iterator, typename OidMap>
+    error_code convert(Iterator&& result, const OidMap& oid_map) {
```
IMO ```in``` or something like this is better than name ```result``` which is also a name of the class.

dpotepalov replied:
done

thed636 commented on include/ozo/binary_deserialization.h:
```diff
@@ -0,0 +1,36 @@
+}
+
+template <typename OidMap, typename T>
+inline error_code convert(oid_t oid, const char* bytes, std::size_t size, const OidMap& oid_map, T& value) {
```
IMO it is better to use ```value``` type or it's concept instead of these three parameters here:
```c++
oid_t oid, const char* bytes, std::size_t size
```

dpotepalov replied:
done

thed636 commented on include/ozo/libpq.h:
```diff
@@ -0,0 +1,143 @@
+        int row_ {0};
+        int col_ {0};
+
+        value v_ {nullptr, 0, 0};
```
The value contains same data elements as iterator itself - it is very strange and redundant.

dpotepalov replied:
done

thed636 commented on include/ozo/libpq.h:
```diff
@@ -0,0 +1,143 @@
+        PGresult * r_;
+        int row_;
+
+        row v_ {nullptr, 0};
```
Same as above.

dpotepalov replied:
done

thed636 commented on include/ozo/libpq.h:
```diff
@@ -0,0 +1,143 @@
+    class iterator : public boost::iterator_facade<
+        iterator,
+        row,
+        boost::forward_traversal_tag, // can be random_access, but we do not need it
```
Why not?

dpotepalov replied:
it's unnecessary for conversion as implemented now, because generic detail::result iterates forwards through libpq::result rows. If we decide to make detail::result (and thus ozo::result) a full-fledged container with random access, then we'll have to implement random access here.

thed636 replied:
IMO there are not so many changes is needed to make a random access iterator

thed636 commented on include/ozo/libpq.h:
```diff
@@ -0,0 +1,143 @@
+
+    // TODO: replace usage of these in impl::row with a random access row iterator
+    std::size_t size() { return PQnfields(r_); }
+    value at(int i) { return value{r_, row_, i}; }
```
What happened here in case of ``` i``` is out of range?

dpotepalov replied:
done - mimicked std:: containers behaviour

thed636 commented on tests/row.cpp:
```diff
@@ -0,0 +1,202 @@
+// TODO: maybe we should try to hide this behind OZO_ADAPT_STRUCT or something?
+// In order to hide usage of fusion or hana as an implementation detail.
+// Would involve a bunch of generated code, though.
+BOOST_FUSION_ADAPT_STRUCT(fusion_adapted, a, b)
```
We shouldn't. It is not the best idea to force a user to use yet another one adaptation system for types and structures. The good choice is to support these two systems simultaneously and ask a user only to provide additional information like type name in a target database.

dpotepalov replied:
ok, comment removed

thed636 commented on include/ozo/impl/result.h:
```diff
@@ -0,0 +1,49 @@
+template <typename Rows, typename OidMap, typename Iterator>
+Require<Iterable<Rows> && (ForwardIterator<Iterator> || InsertIterator<Iterator>), error_code>
+convert_result(Rows&& rows, const OidMap& oid_map, Iterator result) {
+    error_code ec = error::ok;
```
A right variable indicating no error will be:

```c++
error_code ec;
```

Since ```error::ok``` is still an error code. It is just reserved value to not be confused with zero value error code.

dpotepalov replied:
done

thed636 commented on tests/binary_deserialization.cpp:
```diff
@@ -0,0 +1,59 @@
+        const char* bytes = reinterpret_cast<const char*>(&bytes_storage);
+        const auto size = sizeof(int16_t);
+
+        EXPECT_EQ(ok, convert(INT2OID, bytes, size, oid_map, got));
```
You should expect 
```c++
!convert(INT2OID, bytes, size, oid_map, got)
```
here.

dpotepalov replied:
done

elsid commented on tests/binary_deserialization.cpp:
```diff
@@ -0,0 +1,59 @@
+        int16_t got;
+
+        const int16_t bytes_storage = detail::convert_to_big_endian(expected);
+        const char* bytes = reinterpret_cast<const char*>(&bytes_storage);
```
Why perform data conversions in the test? Just define value as it should be.

dpotepalov replied:
done

thed636 commented on include/ozo/libpq.h:
```diff
@@ -0,0 +1,159 @@
+
+namespace ozo::libpq {
+
+class value {
```
IMO no friend class for value is needed at all. Did you try something like proxy object?
```c++
class value {
public:
    struct data { PGresult* res; int row; int col; };
    value(const data& v) : v_(v) {}
//...
private:
    data& v_;
};

//...

class row {
public:
    class iterator : public boost::iterator_facade<
//...
        iterator,
        value,
        boost::forward_traversal_tag, // maybe will implement random access later
        value, // not a reference, but proxy
        int
    > {
        public:
        //...
        private:
            value::data v_;
            value dereference() const {
                return {v_};
            }
//...
```

dpotepalov replied:
I considered such an option and decided that creating an object inside a dereference() method looks a little awkward. If you both decide that it doesn't - ok, proxy it is. There is no need for value::data, actually - you can store res, row and col in row as separate variables and create an instance of value in dereference()

thed636 replied:
--Well, IMO return wrapped reference is better than copy pointer + two integers.--

I'm wrong. The wrapped reference to data is bad way - the only right version is to copy all coordinates into proxy

thed636 commented on include/ozo/libpq.h:
```diff
@@ -0,0 +1,159 @@
+    class iterator : public boost::iterator_facade<
+        iterator,
+        value,
+        boost::forward_traversal_tag, // maybe will implement random access later
```
Why not implement it right now? Any problems?

dpotepalov replied:
Unused code? Three more methods of iterator_facade need to be implemented. Not much work, sure, but that code will be just sitting there doing nothing, potentially requiring maintenance in the future.

thed636 commented on include/ozo/libpq.h:
```diff
@@ -0,0 +1,159 @@
+inline row::iterator end(const row& r) { return r.end(); }
+
+inline result::iterator begin(const result& r) { return r.begin(); }
+inline result::iterator end(const result& r) { return r.end(); }
```
Are these implementations really needed?

dpotepalov replied:
actually not, they were needed at some point. I'll remove them.

Commit:
Simplify serialization a little bit

It is no needs to overload serialization for a single byte
integral value. But it is needed to overload it for bool,
since bool in PG represented by byte.

Commit:
Move floating point to integral staff in separate header

It will be useful in deserealization part.

Commit:
Fix float serialization

Integer representation must be converted due byte ordering rules.

dpotepalov commented on include/ozo/binary_deserialization.h:
```diff
@@ -0,0 +1,157 @@
+inline void recv(const value<T>& in, const oid_map_t<M>& oid_map, Out& out) {
+
+    if (!accepts_oid(oid_map, out, in.oid())) {
+        throw system_error(error::oid_type_mismatch, "unexpected oid "
```
why throw? I thought we decided to use error codes. Although this kind of error does seem unrecoverable.

thed636 replied:
Well, the error_code and context depended information will be provided for user callbacks by means of internal logic and it's arguments. On this level the error_code is not useful because it can not provide context depended information about an error.

dpotepalov commented on include/ozo/binary_deserialization.h:
```diff
@@ -0,0 +1,157 @@
+}
+
+template <typename T, typename M, typename Out>
+Require<InsertIterator<Out>>
```
seems unnecessary, since Out is just basic_result<T> and body does not require InsertIterator support

thed636 replied:
Yep, thanks, it seems like I forgot a unit test for the case.

thed636 replied:
The comment is moved to the different part of code by github - funny :)

dpotepalov commented on include/ozo/impl/io.h:
```diff
@@ -7,6 +7,17 @@
+using result_ptr = std::unique_ptr<PGresult, decltype(&PQclear)>;
+
+/**
+* Here we describe constants for query result format
```
enum name seems self-explanatory, IMO comment is unnecessary here

thed636 replied:
Yep, you right

dpotepalov commented on tests/binary_deserialization.cpp:
```diff
@@ -0,0 +1,244 @@
+    StrictGMock<pg_result_mock> mock{};
+    ::ozo::value<pg_result_mock> value({object(mock), 0, 0});
+
+    SHOULD("return type mismatch error if oid does not match the type") {
```
throw type mismatch error?

thed636 replied:
Yep, thanks.

dpotepalov commented on tests/binary_deserialization.cpp:
```diff
@@ -0,0 +1,244 @@
+    StrictGMock<pg_result_mock> mock{};
+    ::ozo::row<pg_result_mock> row({object(mock), 0, 0});
+
+    SHOULD("return type mismatch error if size of tuple does not equal to row size") {
```
as above - throw type mismatch error?

thed636 replied:
Yep, thanks a lot!
